### DISCUSSION TITLE
Optocoupler input voltage limits

### DISCUSSION THREAD

**Discussion by jakobwal** [hatlabs/discussions#27] [Q&A] [answered]
Hi

Just got my ESP32! I want to use the optocoupler input for an RPM signal from my Yanmar 3GM30. As far as I can tell, this sensor provides an AC signal. Can the ESP32 HAT handle the reverse polarity voltage on the optocoupler input and if so, is it still maximum 18V? Also, (since you have the same engine @mairas :-)) does anyone know it he Yanmar 3GM30 stays within the limits of the ESP32 with regards to voltage and can be directly connected to the optocoupler input or do I need to add in-line resistors?

Thanks
Jakob

**Comment by mairas** [comment, marked as answer]
I don't remember whether the Yanmar tach sensor output actually goes negative, but even if it does, there's a diode for reverse polarity protection:

<img width="500" alt="Screenshot 2022-03-11 at 11 55 09" src="https://user-images.githubusercontent.com/407010/157845030-aa13518f-cb51-4157-8c2a-91a7dfcf8c21.png">

So, electrically it should definitely work.

BUT! I haven't had great success with reading the Yanmar RPM sender with the optocoupler on SH-ESP32. The inductive sensor output is quite weak (in more technical terms, it has a high output impedance), and the relatively low input impedance of the optocoupler input may pull the voltage sufficiently low to make the measurement unreliable.

I am just about to release a new product, the SH-ESP32 Engine Top Hat which is an add-on board for the SH-ESP32, intended for engine and tank measurements. The design repo is here: https://github.com/hatlabs/SH-ESP32-engine-hat. I wanted the engine hat tach inputs to work with a wide variety of sensors, and in the end, I settled with this very simple design:

<img width="562" alt="Screenshot 2022-03-11 at 12 07 15" src="https://user-images.githubusercontent.com/407010/157847267-f09ebdae-5148-42ae-8944-445d3dfa445c.png">

So, basically a large input resistance and two diodes to clamp the voltage between 0V and 3.3V. I've found that to work reliably both with Instead of trying to use the optocoupler, you might want to implement a similar circuit on the proto board area (or order an SH-ESP32 Engine Hat once it finally becomes available!).

**Reply by jakobwal** [reply]
Thanks for a great answer!

Here is some of my own experience with the Yanmar RPM sender. and a follow up question:
About one year ago I built an RPM sensor based on SensESP and a ESP-8266EX board. Used an optocoupler for the input with same components as the ones on SH-ESP32. I don't remember the value for the input resistor but I think it was higher than 2.2k. I based my choice of resistance on some measurements I did with a reasonably dumb multimeter. Did not have access to oscilloscope so I had to assume sine wave when doing my voltage measurements. I got it all to work when I powered my ESP-8266EX through the USB port and my laptop. When I connected it to boat power I got wrong (higher) values. I assumed that there was some ripple on power input that caused the input to "double tap" but I was never sure about this. I was planning to put a low pass filter on the power input to see if that solved the problem but never go around to it.

My follow up question is on the diagram for the Engine Top Hat:

- What is LP1?
- Do you intend to connect one side of the Yanmar RPM sender to the Engine Top Hat GND and the other to InF1?

Thanks again!

**Reply by mairas** [reply]
> * What is LP1?

There's a jumper you can use to enable a LP filter on the input pin to filter out noise. See the full schematics here:

[SH-wg-schematics_1.0.0.pdf](https://github.com/hatlabs/discussions/files/8232379/SH-wg-schematics_1.0.0.pdf)

> * Do you intend to connect one side of the Yanmar RPM sender to the Engine Top Hat GND and the other to InF1?

Yes. Or actually, to In1 - there's an EMI input filter that I omitted in the first screenshot.
